### ISSUE TITLE
tt.descriptor_load falls back to scalar gather instead of 2D block read

### ISSUE THREAD

**Issue by etiotto** [intel/intel-xpu-backend-for-triton#6808]
**Update (2026-05-05)**: the gap has mostly closed on its own. Reran
the benchmark on PVC 1100 today — main uses vectorized `load <4 x i32>`
now, not scalar gather, and hits ~270/446 GB/s at NROWS=8192/65536.
The 2D-block-IO path from PR #6811 comes out neutral-to-slightly-slower
on this workload (267/433 GB/s). Likely an IGC/driver update since
this was filed.

PR #6811 landed as an opt-in pass
(`TRITON_INTEL_FIXUP_DESCRIPTOR_ENCODING=1`) so we can re-measure
if the trade-off shifts again, but it's not default-on.

---

PyTorch Inductor generates Triton kernels that use `tl.make_tensor_descriptor` for memory-intensive operations such as softmax. A representative example is the following fused softmax kernel:

```python
@triton.jit
def triton_red_fused__softmax__unsafe_view_add_clone_permute_4(
    in_ptr0, in_ptr1, out_ptr2, out_ptr3, xnumel, r0_numel,
    XBLOCK: tl.constexpr, R0_BLOCK: tl.constexpr
):
    xoffset = tl.program_id(0) * XBLOCK
    tma_descriptor0 = tl.make_tensor_descriptor(
        in_ptr0,
        shape=[2304, 576, 16],
        strides=[9216, 16, 1],
        block_shape=[(15 + XBLOCK) // 16, R0_BLOCK, 16],
    )
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK):
        tmp0 = tma_descriptor0.load([xoffset // 16, r0_offset, xoffset % 16])
        ...
```

This kernel performs 3 descriptor loads per loop iteration over 576 reduction elements, making memory bandwidth the dominant cost. The Intel hardware supports 2D block read instructions which are significantly more efficient than scalar gathers for this access pattern.

However, `tt.descriptor_load` operations in elementwise and reduction kernels (i.e., kernels that do not involve matrix multiplication) always fall back to scalar gather loads.

The problem is reproducible with a minimal 2D case:

```python
@triton.jit
def load_2d_desc(in_ptr0, out_ptr, ROWS: tl.constexpr, COLS: tl.constexpr):
    pid = tl.program_id(0)
    desc = tl.make_tensor_descriptor(
        in_ptr0, shape=[2304, 16], strides=[16, 1], block_shape=[ROWS, COLS])
    tile = desc.load([pid * ROWS, 0])
    desc_out = tl.make_tensor_descriptor(
        out_ptr, shape=[2304, 16], strides=[16, 1], block_shape=[ROWS, COLS])
    desc_out.store([pid * ROWS, 0], tile)
```

When compiled with `ROWS=64, COLS=16, num_warps=4` on a device with `has_subgroup_2d_block_io=True`, the generated LLIR contains scalar loads rather than 2D block read/write instructions.

**Comment by etiotto** [comment]
## Microbenchmark: 2D Block I/O vs Vectorized Gather (PVC 1100)

Measured the performance impact of lowering `tt.descriptor_load`/`tt.descriptor_store` to 2D block I/O (`LSC2DBlockRead`/`LSC2DBlockWrite`) versus the current vectorized gather path on main.

### Setup
- **Hardware**: PVC 1100 (HBM2e)
- **Kernel**: Contiguous row-major 2D copy via `tl.make_tensor_descriptor` — load + store
- **Tensor**: f16, 256 columns, variable row count
- **Block shape**: 64×256 per program, `num_warps=4`
- **Measurement**: 10 independent `do_bench(warmup=100, rep=200)` samples per size
- **Correctness**: Verified via `torch.allclose` at each size

### LLIR Codegen Verification
- **Fix branch**: LLIR contains `llvm.genx.GenISA.LSC2DBlockRead.v32i16` (4 calls) and `llvm.genx.GenISA.LSC2DBlockWrite.v8i16` (16 calls) — hardware 2D block I/O
- **Main branch**: LLIR contains `getelementptr` + vectorized loads — gather path, no 2D block ops

### Results

| NROWS | Size (MB) | Fix BW (GB/s) | Main BW (GB/s) | Δ% |
|------:|----------:|--------------:|---------------:|----:|
| 8,192 | 8.4 | 282 | 261 | **+8.0%** |
| 65,536 | 67.1 | 441 | 416 | **+6.0%** |

At larger sizes (262K, 1M rows) both paths converge to ~690 GB/s as HBM saturates — the improvement is most visible at small-to-medium tensor sizes where launch overhead and memory access efficiency matter more.

### Conclusion
The 2D block I/O path delivers a **6–8% bandwidth improvement** over vectorized gathers in this contiguous copy microbenchmark. The benefit is expected to be larger for non-contiguous access patterns (column-major, strided, transposed) where gathers degrade significantly.

**Comment by etiotto** [comment]
### Microbenchmark Script

The benchmark used in the experiment above:

```python
import triton
import triton.language as tl
import torch
import numpy as np
from triton.testing import do_bench

@triton.jit
def load_2d_desc(in_ptr0, out_ptr, NROWS: tl.constexpr, NCOLS: tl.constexpr,
                 BLOCK_ROWS: tl.constexpr, BLOCK_COLS: tl.constexpr,
                 STRIDE_ROW: tl.constexpr):
    pid = tl.program_id(0)
    desc = tl.make_tensor_descriptor(
        in_ptr0, shape=[NROWS, NCOLS], strides=[STRIDE_ROW, 1],
        block_shape=[BLOCK_ROWS, BLOCK_COLS])
    tile = desc.load([pid * BLOCK_ROWS, 0])
    desc_out = tl.make_tensor_descriptor(
        out_ptr, shape=[NROWS, NCOLS], strides=[STRIDE_ROW, 1],
        block_shape=[BLOCK_ROWS, BLOCK_COLS])
    desc_out.store([pid * BLOCK_ROWS, 0], tile)

NCOLS = 256
BLOCK_ROWS = 64
BLOCK_COLS = 256
NROWS_LIST = [8192, 65536, 262144, 1048576]

print(f"{'NROWS':>10} {'Size MB':>8} {'Grid':>6} {'Time ms':>10} {'BW GB/s':>10} {'CoV%':>6} {'Correct':>8}")
print("-" * 70)

for NROWS in NROWS_LIST:
    src = torch.randn(NROWS, NCOLS, device='xpu', dtype=torch.float16)
    dst = torch.empty_like(src)
    grid = (NROWS // BLOCK_ROWS,)
    nbytes = 2 * src.numel() * src.element_size()

    for _ in range(50):
        load_2d_desc[grid](src, dst,
                           NROWS=NROWS, NCOLS=NCOLS,
                           BLOCK_ROWS=BLOCK_ROWS, BLOCK_COLS=BLOCK_COLS,
                           STRIDE_ROW=NCOLS, num_warps=4)
    torch.xpu.synchronize()

    times = []
    for i in range(10):
        ms = do_bench(lambda: load_2d_desc[grid](src, dst,
                           NROWS=NROWS, NCOLS=NCOLS,
                           BLOCK_ROWS=BLOCK_ROWS, BLOCK_COLS=BLOCK_COLS,
                           STRIDE_ROW=NCOLS, num_warps=4),
                      warmup=100, rep=200)
        times.append(ms)

    times = np.array(times)
    bws = nbytes / (times * 1e-3) / 1e9
    cov = times.std() / times.mean() * 100

    dst.zero_()
    load_2d_desc[grid](src, dst,
                       NROWS=NROWS, NCOLS=NCOLS,
                       BLOCK_ROWS=BLOCK_ROWS, BLOCK_COLS=BLOCK_COLS,
                       STRIDE_ROW=NCOLS, num_warps=4)
    correct = "PASS" if torch.allclose(src, dst) else "FAIL"

    print(f"{NROWS:>10} {nbytes/1e6:>8.1f} {grid[0]:>6} {times.mean():>10.4f} {bws.mean():>10.2f} {cov:>6.1f} {correct:>8}")

    del src, dst
    torch.xpu.empty_cache()
```

Run with: `/home/jovyan/.conda/bin/python bench_2d_desc.py`

**Comment by etiotto** [comment]
## Design

The root cause is that `tt.descriptor_load` and `tt.descriptor_store` operations receive a blocked encoding from the Coalesce pass that is optimized for pointer-based access patterns (e.g. `sizePerThread=[1, 8]`). This encoding is incompatible with 2D block I/O: MaterializeBlockPointer requires `sizePerThread=[1, 1]` with a row-major order so it can map the tensor tile directly onto hardware 2D block read/write instructions.

The fix adds a new pass, **FixupDescriptorEncoding**, inserted in the TTGIR pipeline immediately after Coalesce and before RemoveLayoutConversions:

```
coalesce → fixup_descriptor_encoding → remove_layout_conversions → ...
```

The pass:
1. Walks the module collecting `DescriptorLoadOp` and `DescriptorStoreOp` with `BlockedEncodingAttr`.
2. Constructs a new `BlockedEncodingAttr` with `sizePerThread=[1,...,1]` and row-major order — the canonical layout for 2D block I/O compatibility.
3. If the new encoding differs from the existing one:
   - **Loads**: rewrites the result type to the new encoding and inserts a `ConvertLayoutOp` after the load to convert back to the original encoding for downstream consumers.
   - **Stores**: inserts a `ConvertLayoutOp` before the store to convert the input from the original encoding to the new one.
4. The subsequent RemoveLayoutConversions pass eliminates any redundant conversions.

The pass is gated on `ttig.support_2d_block_io` — it is a no-op on hardware without 2D block I/O support.

**Comment by etiotto** [comment]
Closing as superseded. Re-measured on PVC 1100 today: main hits ~270/430 GB/s at NROWS=8192/65536, matching the historical "Fix" numbers from PR #6811 without any 2D block IO being emitted — the vectorized gather path got faster on its own. `LowerTo2DBlockLoad` (#6801, #6856) is now the canonical mechanism if/when descriptor → 2D block IO becomes worthwhile for elementwise/reduction kernels. PR #6811 closed.
